### PR TITLE
SetStatusCondition[NoHeartbeat]: return 'changed' bool

### PR DESCRIPTION
**Pull request by 2uasimojo**
Add a boolean return value to SetStatusCondition and
SetStatusConditionNoHeartbeat indicating if the function made any
changes to the conditions (other than the heartbeat timestamp in the
case of the former). This allows consumers to avoid thrashing by
requeueing an object that has not changed substantively.

This represents an additive API change which should be backward
compatible as existing consumers will simply be ignoring the new return
value.

### CODE REVIEW

**Comment by 2uasimojo** [comment]
/assign @djzager 

FYI, we're interested in [using this from hive](https://github.com/openshift/hive/pull/1672#discussion_r860290708).

File: conditions/v1/conditions_test.go [hunk 1]
```diff
@@ -113,20 +117,57 @@ func TestSetStatusCondition(t *testing.T) {
...
 		},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			SetStatusCondition(tc.startConditions, tc.testCondition)
-			compareConditions(t, tc.startConditions, tc.expectedConditions)
+			// Copy tc.startConditions so it doesn't get updated in place
+			startConditions := make([]Condition, len(*tc.startConditions))
+			copy(startConditions, *tc.startConditions)
```
Comment (2uasimojo) [inline conditions/v1/conditions_test.go:154]: note latent bug in the test driver: previously the first `SetStatusCondition` was overwriting `tc.startConditions` so it was already in the expected state when it got to `SetStatusConditionNoHeartbeat`.

File: conditions/v1/conditions_test.go [hunk 2]
```diff
@@ -113,20 +117,57 @@ func TestSetStatusCondition(t *testing.T) {
...
+			changed = SetStatusConditionNoHeartbeat(&startConditions, tc.testCondition)
+			if changed != tc.expectChanged {
+				t.Errorf("Unexpected return from SetStatusConditionNoHeartbeat: expected: %t; actual: %t", tc.expectChanged, changed)
+			}
+			compareConditionsNoHeartbeat(t, &startConditions, tc.expectedConditions)
 		})
 	}
-
-	return
```
Comment (2uasimojo) [inline conditions/v1/conditions_test.go:129]: unrelated fix: VSCode linter complained of redundant return.

File: conditions/v1/conditions_test.go [hunk 3]
```diff
@@ -191,8 +232,6 @@ func TestRemoveStatusCondition(t *testing.T) {
 			compareConditions(t, tc.startConditions, tc.expectedConditions)
 		})
 	}
-
-	return
```
Comment (2uasimojo) [inline conditions/v1/conditions_test.go:195]: unrelated fix: VSCode linter complained of redundant return.

**Comment by djzager** [comment]
@2uasimojo isn't the only substantive change in `SetStatusCondition`, when called repeatedly, the `LastHeartBeatTime`? Could you expand some on what this gives you that beyond `SetStatusConditionNoHeartbeat`?

I don't see anything technically wrong but some more information about the problem being addressed would be greatly appreciated

**Comment by 2uasimojo** [comment]
@djzager Sure thing. I tried to explain it in the commit message, but was pretty brief about it :)

The code example in [conditions/README.md](https://github.com/openshift/custom-resource-status/blob/a29288cd97c4a42bc143146b17002dae7c9078b7/conditions/README.md) (before this change) looks roughly like:

```go
func (...) Reconcile(...) .. {
  conditions.SetStatusCondition(...)
  r.client.Status().Update(...)
}
```

Assuming this `Reconcile` is like most, the controller manager is `Watch()`ing for changes to the object. So the `Status().Update()` will enqueue the object to be reconciled again. And then we'll `SetStatusCondition()` again, which will update nothing... except the heartbeat. And then `Status().Update()` will trigger the reconcile again. Etc. -- hot loop.

Now, arguably if your controller is designed this way, you shouldn't be using heartbeats at all. This is probably what Hive will do. But let's say you want to reconcile immediately on "normal" changes, but still heartbeat every 2m. You could do something like:

```go
const (
  heartbeatInterval = 2 * time.Minute
)
...
func (...) Reconcile(...) .. {
  changed := conditions.SetStatusCondition(...)
  timeSinceLastHeartbeat := time.Now().Sub(cond.LastHeartbeatTime)
  if changed || timeSinceLastHeartbeat >= heartbeatInterval {
    r.client.Status().Update(...)
    // If only heartbeating, delay requeue
    requeueAfter := 0
    if !changed {
      requeueAfter = timeSinceLastHeartbeat
    }
    return reconcile.Result{RequeueAfter: requeueAfter}, nil
  }
  ...
}
```

...but if you don't need anything like that, you can just ignore the return value from `SetStatusCondition()` :)

**Comment by djzager** [comment]
@2uasimojo I really appreciate the contribution. The part where I'm struggling is in understanding what this change buys users of this library that `SetStatusConditionNoHeartbeat` does not.

As I think through this specific scenario,
> Assuming this Reconcile is like most, the controller manager is Watch()ing for changes to the object. So the Status().Update() will enqueue the object to be reconciled again. And then we'll SetStatusCondition() again, which will update nothing... except the heartbeat. And then Status().Update() will trigger the reconcile again. Etc. -- hot loop.Now, arguably if your controller is designed this way, you shouldn't be using heartbeats at all. This is probably what Hive will do. But let's say you want to reconcile immediately on "normal" changes, but still heartbeat every 2m. 

There are two possibilities that I am currently aware of:

1. Use `SetStatusConditionNoHeartbeat`.
2. Use [predicates](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/predicate/predicate.go)

For the other scenario you mentioned:
> Now, arguably if your controller is designed this way, you shouldn't be using heartbeats at all. This is probably what Hive will do. But let's say you want to reconcile immediately on "normal" changes, but still heartbeat every 2m. 

From my perspective, the `lastHeartbeatTime` loses it's value if it isn't being updated every time the condition is observed (or calculated). You could potentially call `SetStatusCondition` and manage the `lastHeartbeatTime` yourself.

**Comment by djzager** [comment]
One thing I wonder @2uasimojo is if we could implement a whole new function like `SetStatusConditionHeartbeat(conditions *[]Condition, newCondition Condition, heartbeatFn HeartbeatFunction)` that accepted a function holding the logic for whether or not to change the `lastHeartbeatTime`? The idea being, if there is no change to the condition, this `heartbeatFn` would be called (returning a boolean) to determine if the `lastHeartbeatTime` should be updated. In the event something else about the condition changed, the `lastHearbeatTime` would be updated.


That would make it so we have very clear answers to the different use cases:
1. If you want heartbeat everytime but don't want to ddos your reconciler, you should consider using [the `GenerationChagedPredicate`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/predicate#GenerationChangedPredicate) 
1. If you don't care about heartbeat, use `SetStatusConditionNoHeartbeat`.
1. If you want it, but on an interval, use this new function tentatively named `SetStatusConditionHeartbeat(...)`

**Comment by 2uasimojo** [comment]
Thanks for the thoughtful responses @djzager 

Honestly, I don't really care about the change to `SetStatusCondition` -- I was mainly doing it for symmetry. The one I care about, and the one hive will use exclusively (because we don't use heartbeats) is `SetStatusConditionNoHeartbeat`. Do you agree the boolean return from that one is generally useful without stretching for scenarios? If so I can just revert the other bit and we're good to go.

**Review by djzager — APPROVED** [review]
I too prefer symmetry. While I don't see clearly how this benefits consumers of this library, in the end my calculus looks like:

1. I'm not able to articulate any reason why this change shouldn't be implemented. While I have some concerns about how this could be useful, I don't believe that is sufficient for a NACK.
2. The change is a no-op for current consumers of the library.
3. The change improves the state of code written and tests.

For those reasons I think it should be accepted. Thanks for your patience @2uasimojo.

/lgtm

**Comment by 2uasimojo** [comment]
Thanks @djzager! Here's an [example](https://github.com/openshift/hive/pull/1672#discussion_r864062408) of how we intend to use this.
